### PR TITLE
add apisix-version info to headers

### PR DESCRIPTION
**Pull request by swayamraina**
### What this PR does / why we need it:
We should pass the APISIX version in the response headers
fixes : https://github.com/apache/incubator-apisix/issues/1877

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [ ] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [ ] Is this PR backward compatible?

### CODE REVIEW

**Comment by membphis** [comment]
please take a look at this patch:

```diff
diff --git a/apisix/init.lua b/apisix/init.lua
index 354290dc..e465a372 100644
--- a/apisix/init.lua
+++ b/apisix/init.lua
@@ -39,6 +39,7 @@ local load_balancer
 local local_conf
 local dns_resolver
 local lru_resolved_domain
+local apisix_ver    = "APISIX/" .. core.version.VERSION


 local function parse_args(args)
@@ -299,6 +300,8 @@ function _M.http_access_phase()

     core.ctx.set_vars_meta(api_ctx)

+    core.response.set_header("server", apisix_ver)
+
     -- load and run global rule
     if router.global_rules and router.global_rules.values
        and #router.global_rules.values > 0 then
```

**Comment by swayamraina** [comment]
After reading more code and integration with `nginx.conf`, I guess this header should be added during the init phase and not in the response phase to cater to error scenarios as well.

**Comment by membphis** [comment]
> After reading more code and integration with `nginx.conf`, I guess this header should be added during the init phase and not in the response phase to cater to error scenarios as well.

the nginx can not set the response header in the `init` phase.

**Comment by swayamraina** [comment]
@membphis  Yes, I meant the access phase only. 
I see you recommend creating a local variable and then do an append operation for every incoming request. 
Though I think we should create a global constant in `version.lua` and use it directly to avoid using some CPU in every request?

**Comment by nic-chen** [comment]
> Yes, I meant the access phase only.

I think we should add apisix-version in `bin/apisix`.
If dynamically modify headers, it may affect performance, doesn't it ?

**Comment by moonming** [comment]
> > Yes, I meant the access phase only.
> 
> I think we should add apisix-version in `bin/apisix`.
> If dynamically modify headers, it may affect performance, doesn't it ?

do you means `more_set_headers` directive? I think it's ok to do this job in Lua side, and maybe more heads will be added later.

File: apisix/init.lua [hunk 1]
```diff
@@ -299,6 +300,8 @@ function _M.http_access_phase()
 
     core.ctx.set_vars_meta(api_ctx)
 
+    core.response.set_header("server", ver_header)
```
Comment (juzhiyuan) [inline apisix/init.lua:303]: I would recommend adding a more independent header to describe the version of APISIX, e.g {"X-Version": "1.4"}

Comment (membphis) [reply apisix/init.lua:303]: The current way is more common, we should use the same way.

* tengine: `Server: Tengine/2.2.3`
* kong: `Server: kong/1.4.0`
* nginx: `Server: nginx/1.14.2`

Comment (membphis) [inline apisix/init.lua:303]: use `"Server"` is easier to read.

Comment (swayamraina) [reply apisix/init.lua:303]: Agreed. Moreover `X-Version : 1.4` does not convey any information related to `apisix`

Comment (swayamraina) [reply apisix/init.lua:303]: Sure.

**Comment by membphis** [comment]
@swayamraina you need to update the test cases.

CI: https://github.com/apache/apisix/pull/1961/checks?check_run_id=950677448#step:6:280

```
# received: Connection: close
# received: Server: APISIX/1.4.1
# received: Server: Tengine
# received: 
# received: hello world
# close: 1 nil
# '
#     doesn't match '(?^:connected: 1
```

**Comment by swayamraina** [comment]
> Can you please help to understand how to setup/run TCs (from https://github.com/apache/apisix/issues/1877)

@membphis  I am unable to run TCs locally. It produces the following error.

> Can't locate Test/Nginx/Socket/Lua/Stream.pm in INC (you may need to install the Test::Nginx::Socket::Lua::Stream module)
> BEGIN failed--compilation aborted at t/APISIX.pm line 21.

**Comment by membphis** [comment]
you can take a look at the CI install script: https://github.com/apache/apisix/blob/master/.travis/linux_openresty_runner.sh

another doc: https://github.com/apache/apisix/blob/master/doc/how-to-build.md#4-test

File: t/node/remote-addr-ipv6.t [hunk 2]
```diff
@@ -114,6 +114,7 @@ request sent: 59
 received: HTTP/1.1 200 OK
 received: Content-Type: text/plain
 received: Connection: close
+received: Server: APISIX/1.4.1
```
Comment (membphis) [inline t/node/remote-addr-ipv6.t:117]: the version is changing always.
We need to use regular matching.

Comment (swayamraina) [reply t/node/remote-addr-ipv6.t:117]: Yeah agreed. We do not want to update TCs every time we do a version bump up.
Wanted to check if the pipeline is a success or not.

File: t/router/radixtree-sni.t [hunk 3]
```diff
@@ -313,6 +314,7 @@ sent http request: 62 bytes.
 received: HTTP/1.1 200 OK
 received: Content-Type: text/plain
 received: Connection: close
+received: Server: APISIX/(\d+)\.(\d+)\.(\d+)
```
Comment (membphis) [inline t/router/radixtree-sni.t:317]: `(\d+)\.(\d+)\.(\d+)` -> `([\w\.]+)`

**Comment by moonming** [comment]
why?  The version can only be number

YuanSheng Wang <notifications@github.com> 于 2020年8月8日周六 下午5:42写道：

**Comment by membphis** [comment]
> why? The version can only be number

There are two version numbers for the version number, such as `1.4` or `1.4.1` now. 
`([\d\.]+)` is better pattern.

File: t/router/radixtree-sni.t (hunk 3 above)
Comment (membphis) [reply t/router/radixtree-sni.t:317]: There are two version numbers for the version number, such as 1.4 or 1.4.1 now.

Comment (swayamraina) [reply t/router/radixtree-sni.t:317]: Should we update this to `([\d\.]+)(\d)$`
`1.4.1fd` and `1.4.1.` also passes `([\d\.]+)`

File: t/node/remote-addr-ipv6.t [hunk 4]
```diff
@@ -114,7 +114,7 @@ request sent: 59
 received: HTTP/1.1 200 OK
 received: Content-Type: text/plain
 received: Connection: close
-received: Server: APISIX/(\d+)\.(\d+)\.(\d+)
+received: Server: APISIX/([\d\.]+)(\d)$
```
Comment (membphis) [inline t/node/remote-addr-ipv6.t:117]: `([\d\.])+` is enough

**Comment by moonming** [comment]
I don't think so.
We need strict restrictions, which is the meaning of the test case.
how about `\d\.\d+(\.\d+)?$`?

Thanks,
Ming Wen
Twitter: _WenMing


YuanSheng Wang <notifications@github.com> 于2020年8月10日周一 上午9:17写道：

**Comment by swayamraina** [comment]
> how about `\d\.\d+(\.\d+)?$`?

Makes sense. This perfectly captures the version correctness.

**Comment by membphis** [comment]
you can make a try with `APISIX/\d\.\d+(\.\d+)?` .

some test cases failed now.

**Pull request merged**